docs/architecture/adr-021-subspace-specific-custom-fee-tokens.md

```diff
@@ -0,0 +1,152 @@
+# ADR 021: Subspace specific custom fee tokens
+
+## Changelog
+
+- April 24th, 2023: First draft;
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR introduces a mechanism that allows users to pay fees within a designated subspace using their existing service tokens.
+
+## Context
+
+Desmos uses subspaces to represent virtual applications built on top of our protocol. Currently, subspaces are mainly used to store content such as posts, reports, and reactions. While this is sufficient for developing DApps on Desmos, it may not be enough to convince existing applications to migrate to our platform since Desmos is lack of a mechanism that enables users to pay fees with existing service tokens.
+
+For instance, consider Keplr's on-chain governance discussions feature, which is currently implemented using a custom backend that stores data in a centralized way. Although Keplr could easily use Desmos as their backend, they are not willing to do so because it offers no advantage to them. Migrating to Desmos would simply transfer their existing users to our chain without any benefits for them.
+
+## Decision
+
+To address the issue mentioned above, we propose implementing a mechanism that allows users to pay fees using the existing service tokens within the specified subspace. The mechanism must have the following features:
+- Defining a minimum gas prices for non-DSM tokens that validators are willing to accept via governance proposal;
+- Allowing the subspace manager(s) to update the list of fee tokens allowed in the subspace;
+- Checking the on-chain minimum gas price requirement for non-DSM fees using a custom `TxChecker`.
+
+The upcoming changes are as follows:
+
+### Types
+
+We will define a new `Params` type to show the `x/subspaces` parameters.
+
+```proto
+// Params contains the module parameters
+message Params {
+  // List of the minimum gas prices to be accepted by validators
+  repeated Coin min_gas_prices = 2;
+}
+```
+
+In addition, we will add a __allowed_fee_tokens__ field to current `Subspace` structure.
+
+```proto
+message Subspace {
+  // Unique id that identifies the subspace
+  uint64 id = 1;
+
+  // Human-readable name of the subspace
+  string name = 2 [ (gogoproto.moretags) = "yaml:\"name\"" ];
+
+  ...skip
+
+  // the creation time of the subspace
+  google.protobuf.Timestamp creation_time = 7;
+  
+  // List of allowed fee tokens inside the subspace
+  repeated string allowed_fee_tokens = 8;
+}
+```
+
+### Msgs
+
+We will allow the the following operations:
+1. update subspace allowed fee tokens list by managers;
+2. update `x/subspaces` parameters by governance.
```

Comment:
Why is this needed? It would be much simpler to require subspace owners to go through a governance-gated update of the subspace `allowed_fee_tokens` list instead of having two different message types for this. Just: 

1. The subspace owner creates an on-chain proposal to update the list of tokens 
2. The governance votes yes/no 
3. Based on the result, the list is updated or not.

No need for multiple message types or anything like that

Reply:
Splitting the message into two parts would make it easier for validators to manage min-gas-prices. 

For example, let's say Subspace A and B both allow a min-gas-price of 0.1 minttoken. If validators want to update the mini-gas-price for minttoken to 0.2 minttoken after a market crash, they would need to perform the following operations:
1. Propose a new min-gas-price of 0.2 minttoken for Subspace A
2. Propose a new min-gas-price of 0.2 minttoken for Subspace B
3. Propose multiple proposals to update the min-gas-price for all subspaces that allow minttoken

However, this approach can make it difficult for validators to manage min-gas-prices. To address this, I decided to create a separate list for minimum gas prices.

Reply:
@dadamu This is extremely complicated to understand and wrap my head around. I honestly cannot understand: 

1. why there's the need to have two different messages.  
   If both subspace managers *and* any other user must use *the same* governance proposal to update the subspace fee tokens, why do we need to have two different messages? 
2. why can't we use an implementation of the `Content` interface, rather than custom messages?  
   Why can't we have something as simple as other governance proposal types such as `SpendCommunityProposal` but instead we need to have this custom message? 


I honestly don't think it's necessary to have this extremely complicated mechanism with two different messages types. Instead, we can simply use a single `Content` implementation type and rely on that

Reply:
@RiccardoM `Content` interface is gov v1beta1 feature, it will be deprecated in the future. Now, cosmos-sdk applies proposal message to have spend community proposal. See:
https://github.com/cosmos/cosmos-sdk/blob/v0.47.2/proto/cosmos/distribution/v1beta1/tx.proto#L155

Reply:
@dadamu Ok, I see what you mean about the `Content` interface. 

For the other discussion, about two messages, it makes little to no sense to have such an incredible complex procedure. Instead, the procedure should be as follow: 

1. The subspace admin creates a governance proposal asking validators if it's OK for them to accept fees paid in the `TokenA` inside their subspace.
2. If the proposal passes, the list of fee tokens **only within that subspace** is updated. 

This guarantees that: 

1. Subspace admins **always** have to ask the governance its permission to accept a new token 
2. The governance can agree to accept `TokenA` as a fee token inside one subspace, but reject it for another subspace (for whatever reason)
3. There is a lower chance of introducing bugs: with a simplified procedure and less operations, fewer things can go wrong

To do this, we can simply implement a `MsgUpdateSubspaceFeeTokens` that accepts: 
- a `SubspaceID`
- a list of coin denominations 

It should be executed only by the governance module, making it de-facto as much similar as the old `Content` interface.

Reply:
@RiccardoM I see what you want to do is reduce the potential bugs, while this procedure is lack of the minimum gas prices management from validators since only Subspace `admin` can create a proposal to modify the minimum gas prices of subspaces. I believe validators would like to fix the minimum gas prices after market crash.

For example, multiple subspaces passed proposals to use stable coin `USDM`. `USDM` depegs and become 0.001USD after market crash. Then validator must ask all the subspaces managers to modify their own fee token list to remove `USDM`, or validators will keep receiving valueless `USDM` as fees.

Can you give me a sample that as a validator how to modify the minimum prices of `USDM` inside each subspace in a one proposal?

Reply:
@dadamu If validators want to modify the minimum gas prices they accept, they should be able to do it by editing the `min-gas-price` inside their node configuration. We could simply read that file and if the value written there is less than the value inside the subspace, we should make sure the transaction is not added to the validator's mem pool. 

This way: 
- subspace admins are the only ones that can touch the fee denoms accepted within their subspace;
- if validators want to independently update their min gas prices, they can do that without a problem;
- if no validator sets a min-gas-price for a specific fee denom, then the default one for that subspace will be used.

Do you think that's technically possible?

Reply:
Yes, it is doable and solves my concerns. I will update ADR to include everything we discussed.

Reply:
@RiccardoM Updated.